XMonad/Hooks/DynamicLog.hs

```diff
@@ -746,6 +764,196 @@ xmobarStripTags tags = strip [] where
+-- >              >> dynamicLogWithPP ppBottom { ppOutput = hPutStrLn xmproc1 }
+-- >              >> dynamicLogWithPP pp1 { ppOutput = hPutStrLn xmproc2 }
+-- >     ...
+-- >   }
+--
+-- Which has a lot of boilerplate. Using the provided functions in this section, the amount of
+-- boilerplate is reduced.
+-- If you wish to have multiple status bars managed by xmonad, you can do so by creating
+-- the status bar configs and combining them with 'combineSBs':
+--
+-- > xmobarTop    = statusBarHandleConfig "xmobar -x 0 $HOME/.config/xmobar/xmobarrc_top"    ppTop
+-- > xmobarBottom = statusBarHandleConfig "xmobar -x 0 $HOME/.config/xmobar/xmobarrc_bottom" ppBottom
+-- > xmobar1      = statusBarHandleConfig "xmobar -x 1 $HOME/.config/xmobar/xmobarrc1"       pp1
+-- > main = xmonad =<< combineSBs [xmobarTop, xmobarBottom, xmobar1] myConfig
+--
+-- Or if you're feeling adventurous:
+--
+-- > myBars = map (uncurry statusBarHandleConfig) [ ("xmobar -x 0 $HOME/.config/xmobar/xmobarrc_top",    ppTop)
+-- >                                              , ("xmobar -x 0 $HOME/.config/xmobar/xmobarrc_bottom", ppBottom)
+-- >                                              , ("xmobar -x 1 $HOME/.config/xmobar/xmobarrc1",       pp1) ]
+-- > main = xmonad =<< combineSBs myBars myConfig
+--
+-- To change the default keybinding or action to toggle struts, you can do the following:
+--
+-- > ...
+-- > mySBs = withAction myAction . combineSBConfs <$> sequence myBars
+-- > main = xmonad =<< makeStatusBar myConfig =<< mySBs
+--
+-- if you also want to change the key, you can chain the transformations like so:
+--
+-- > ...
+-- > mySBCombinator = withKey myToggleStrutsKey . withAction myAction . combineSBConfs <$> sequence myBars
+-- > ...
+--
+-- The above examples also work if the different status bars support different
+-- logging methods: you could do mix property logging and logging via standard input.
+-- One thing to keep in mind: if multiple bars read from the same property, their content
+-- will be the same. If you want to use property-based logging with multiple bars,
+-- they should read from different properties.
+
+-- | Reduces a list of (pure) 'StatusBarConfig's into a single one, by sequencing
+-- their respective hooks. To avoid ambiguitiy, the key used to toggle struts and
+-- the action performed by it are the defaults. If multiple status bars with different
+-- properties are given, make sure they read from different properties.
+-- To directly combine confs created by 'statusBarHandleConfig', 'statusBarHandleConfig''
+-- or 'statusBarPropConfig', take a look at 'combineSBs' instead,
+-- which takes instead a list of 'StatusBarConfig's that are wrapped in 'IO'.
+combineSBConfs :: [StatusBarConfig] -- ^ The list of configs to be merged
+               -> StatusBarConfig
+combineSBConfs confs = def
+  { sbLogHook     = mapM_ sbLogHook confs
+  , sbStartupHook = mapM_ sbStartupHook confs
+  , sbCleanupHook = mapM_ sbCleanupHook confs
+  }
```

Comment:
It'd be a bit nicer if StatusBarConfig was a Monoid or Semigroup. We wouldn't need this then.
But it can't be, because of keyToggle and toggleAction. :-(

Frankly, the entire key/action handling seems a bit weird to me. Like, not just the stuff that you added, even the stuff that was already there. That `XConfig → …` stuff that's there just to let the user override the mod key… Seems like a lot of complexity for little benefit. And then you have keyToggle and toggleAction in StatusBarConfig which, at first glance, appears to possibly be useful for toggling individual bars, but then a comment deep in the docs reveals that no, it doesn't do that. :-/

If we can figure out an easier interface for this, I'd say we can break compatibility. Yeah, we should try to keep the basic simple `dzen`, `xmobar` and `xmobarProp` functions as they are including picking up modkey changes, but the advanced interface doesn't need to be dragged down by this.

(We don't need to block the merge because of this, but if we managed to find some good design before the release, that would be super awesome. :-))

Reply:
Oh and of course we can go the other way. Adjust the types so that StatusBarConfig can be a Semigroup and toggling of individual statusbar is possible. Not sure if that's worth it, but it's definitely possible.

But I should also say that I don't really have a horse in this race:

My setup is still so custom and complicated that it doesn't fit even this new improved interface. I have one xmobar on the primary screen which uses dynamicLogString and xmonadPropLog, but doesn't use xmobarProp/statusBarPropTo/makeStatusBar precisely because I don't want my keybindings touched, and also because I need a more robust version of cleanupStatusBars (parallel killing of multiple xmobars and a trayer, with retries when they refuse to die, implemented using async, which needs threaded runtime and therefore can't ever go to xmonad-contrib, or maybe it could with some ifdefs or runtime detection or something, dunno). Then I have one xmobar per screen (and I change the number of screens very often) that doesn't use dynamicLogString at all. And an event hook on top that keeps the trayer above xmobar.

Therefore I have no strong opinion about this. Just suggestions. ;-)

Reply:
Before this PR I didn't use the `statusBar` for a similar reason; I found it weird that it was required to provide a keybinding that I don't have full control on its functionality. I also initially thought about implementing this as `Monoid` but the `keyToggle/toggleAction`  stuff was in the way.

One suggestion would be to change the interface of `StatusBarConf` so that instead of `keyToggle/toggleAction` the users would need to provide a `(XConfig Layout -> Map (ButtonMask, KeySym) (X ()))` and simply compose them when merging multiple status bars, instead of requiring a top level key/action pair. With some helper functions to make creating these maps easier, we could deprecate `statusBar` and friends in favor of the configs.

Reply:
Regarding your config, you could create multiple `StatusBarConf`s and compose them, and override the `sbCleanupHook` to whatever you want. The problem would be that it's still not enough to react to dynamic screen changes I guess.  I'm planning to move on to `DynamicBars` after we decide on the interface here, and either rewrite it in terms of `StatusBarConf` or leave it and write a similar module. I thought about adding an event hook to the `StatusBarConf`, but I think it's gonna be left unused in `DynamicLog` and only be used in `DynamicBars`, so I didn't

Reply:
Yeah, `XConfig Layout -> Map (ButtonMask, KeySym) (X ())` is definitely better as it allows `const mempty`, i.e. no touching the (key)bindings at all.

What if someone wants to bind their mouse button 9 to toggle the struts? Then the best type is probably `XConfig l → XConfig l`. Or maybe at that point they should go with `const mempty` and add it to the rest of their custom keybindings.

So anyway, let's continue with that thought for a bit. All of StatusBarConfig can be just `XConfig l → XConfig l`. (Yeah, I know that's where we started, and I know I was the one who suggested records. Let's ignore that now.) The only reason we bother with the record in the first place is because we want to make it easier for people to change the keybinding and keep all the other defaults. Other customizations don't benefit from the StatusBarConfig abstraction much, I think, and it's often easier to just do all the plumbing by hand. Or… is it expected that someone would modify other fields of the StatusBarConfig record?

What I'm trying to say here is that we should explicitly think about and agree what usecases we're trying to make easier, what customizations we are trying to abstract over, etc. (And I do have to admit I got lost a bit. Maybe it was explicit somewhere and I missed it. Can you remind me if that is so?)

Reply:
I think the main benefit of `StatusBarConfig` is that we can have composable bars. Whether we include the keybindings or not, as well as the layout modifier, is not really important. I only added it because that functionality was provided in `statusBar` and its variants, and I wanted to provide a replacement with the same functionality. 

I'm kinda tempted to make this another module that is a more user-friendly and more flexible version of `X.H.DynamicBars`

Reply:
Keybindings and layout modifier make composability harder, so I'd say it's best to keep those somewhat separate. A version of `makeStatusBar` that only handles startup/cleanup and logging, but not keybindings/layout would probably be useful, and then maybe another that adds also the rest.

That thing with alternating startup/cleanup hooks leading to only the last status bar staying up makes me think that the composability here is not perfect yet. Right now `sbCleanupHook` is either mempty or `cleanupStatusBars`, and chaining multiple `cleanupStatusBars` is functionally equivalent to having just one `cleanupStatusBars`. Also, `cleanupStatusBars` is noop if nothing was started via `spawnStatusBarAndRemember`, so we might treat `cleanupStatusBars` the same as the layout: it should be there somewhere, but only once, doesn't need to compose with the rest. Therefore `StatusBarConfig` is now just a loghook and a startuphook, everything composes nicely, and it can be just an `XConfig → XConfig` or `XConfig → IO XConfig`. Not saying it should be, but it could be, and it's good to think about it this way, and then composability comes much more naturally. We might prefer a newtype, to prevent accidents, or keep the record so that we know what's what.

Reply:
(Making this a separate module means we wouldn't need to bother with backward compat, but I wouldn't give up just yet. If the interface is good enough to eventually replace the implementation of the most simple API in DynamicLog, everyone benefits.)

Reply:
I see your point, and I agree with separating the keybindings and the layout from the rest. 

However, while `sbCleanupHook` might not really look useful here, I was thinking about using it later to provide functionality similar to `DynamicBars`, by triggering a cleanup and a restart as a reaction to a change in the screen configuration. In that case, `sbCleanupHook` would probably be either a `cleanupStatusBars` or closing a handle, and chaining multiple cleanup hooks might be necessary since you really don't know what to do to each bar. I also think that using a record has the slight advantage that it might be easier for users to go beyond the defaults, and provide their own `sb*Hook`s

Reply:
I see, that is a reasonable path forward.

Still, with bar-specific cleanups, the "alternating startup/cleanup hooks leading to only the last status bar staying up" problem doesn't exist. That also means that the ordering between startup and cleanup hooks doesn't really matter, and the startup hook can do the cleanup first. :-)

`cleanupStatusBars` as it is now doesn't fit into that unfortunately. It's nice and simple, though, so maybe what I'm saying is worse. Not sure, really. Maybe we can extend `StatusBarPIDs` with tags (e.g. using [Unique](https://hackage.haskell.org/package/base-4.14.1.0/docs/Data-Unique.html#t:Unique)) so that we can optionally selectively restart individual bars.

About users going beyond the default, I am somewhat sceptical about what value they'd gain by using the record instead of just doing all the plumbing themselves.

Reply:
Not sure if using `Unique` would work though, persisting state between the restarts requires a `Show` instance and `Unique` doesn't seem to provide that.

Reply:
You're right, that wouldn't work. There's certainly some other way to make the tagging work. No need to worry about that now. :-)

Reply:
A simple workaround would be to have the cleanup only kills:

- `partialCleanup` when the screen configuration changes
- `fullCleanup` when restarting xmonad

this makes sense for both property-based and stdin-based loggers, as in the case of a partial cleanup, we can't kill stdin-based loggers with the current setup. I gave this some thought, and we probably can't get away with only one cleanup hook